AK/PrintfImplementation.h

```diff
@@ -283,166 +285,227 @@ ALWAYS_INLINE int print_signed_number(PutChFunc putch, char*& bufptr, int number
     return print_number(putch, bufptr, number, left_pad, zero_pad, field_width);
 }
 
-template<typename PutChFunc>
-ALWAYS_INLINE int printf_internal(PutChFunc putch, char* buffer, const char*& fmt, va_list ap)
+struct State {
+    bool left_pad = false;
+    bool zero_pad = false;
+    bool dot = false;
+    unsigned field_width = 0;
+    bool has_fraction_length = false;
+    unsigned fraction_length = 6;
+    unsigned long_qualifiers = 0;
+    bool size_qualifier = false;
+    bool alternate_form = 0;
+    bool always_sign = false;
+};
+
+template<typename PutChFunc, typename ArgumentListT, template<typename T, typename U = ArgumentListT> typename NextArgument>
+struct PrintfImpl {
+    ALWAYS_INLINE int format_s(PutChFunc putch, char*& bufptr, int, const State& state, ArgumentListT ap) const
```

Comment:
Argument list is passed by value here and by reference below. Why?

Reply:
It's not a reference here, the name is misleading, I suppose.
I'll rename it to `ArgumentListRefT`